### PR TITLE
Add BinaryBackend and BinaryCircuit, update simulator functionalities

### PR DESCRIPTION
**Pull request by paolodr98**
This PR adds the `BinaryCircuit` and `BinaryBackend`, which enable the use of non-linear topologies in the qubit simulation layout. The documentation has been updated, and tests for these new classes have been created.

The BinaryBackend uses an optimization algorithm to reduce the length of the gate list generated by the circuit class. Tests for this optimization are also included.

Additionally, it is now possible to measure fewer qubits than those used in the calculation, thanks to a new function integrated into the simulator.

The main `README.md` has been updated to introduce these new features.

In docs/tutorials/notebooks/, all outdated notebooks have been removed, and an interactive notebook has been added for a Quickstart guide to the simulator, covering both the old and new functionalities.

### CODE REVIEW

File: src/quantum_gates/_utility/circ_optimizator.py [hunk 1]
```diff
@@ -0,0 +1,491 @@
+"""
+This file implement a class that represents an optimization algorithm for the output of the circuit class.
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/circ_optimizator.py:2]: @paolodr98 : Question: How did we come up with the optimizations? Are they just necessary to speed up the execution or do they come from a paper or our own ideas? 

At first, I thought the optimization level might be a Qiskit concept, so you might want to give some background information.

File: src/quantum_gates/_utility/circ_optimizator.py [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+                        proces_snip.append(snip[i])
+                    proces_snip.append([gate,qubits])
+                else:
+                    proces_snip.append([gate,qubits])
+            else: # the closer gate is neither compatible with the targer nor the control qubit
+                #print("before 2.3")
+                for i in range(loc+1):
+                    proces_snip.append(snip[i])
+        else: # if there aren't items before the 2q qubit just append the item
+            #print("before 3")
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/circ_optimizator.py:410]: @paolodr98 Please remove unnecessary comments and unused print statements.

File: src/quantum_gates/_utility/simulations_utility.py [hunk 3]
```diff
@@ -215,3 +218,64 @@ def post_process_split(source_filenames: list, target_filenames: list, split: in
...
+        scheduling_method='asap',
+        seed_transpiler=seed
+        )
+
+        return t_circ
+
+
+    
+
+
```
Comment (romanwixinger) [inline src/quantum_gates/_utility/simulations_utility.py:280]: @paolodr98 Please remove the unnecessary whitespaces. There should be a single newline at the end of each file. Following conventions like this make sure that we do not have our commits polluted with changes that do not change the logic (non-changes like whitespaces or newlines).

File: src/quantum_gates/utilities.py [hunk 4]
```diff
@@ -9,4 +9,8 @@
     setup_backend,
     post_process_split,
     compute_Hellinger_distance as hellinger_distance,
+    create_random_quantum_circuit,
+    transpile_qiskit_circuit
 )
+
+from ._utility.circ_optimizator import Optimizator
```
Comment (romanwixinger) [inline src/quantum_gates/utilities.py:16]: @paolodr98 Maybe we should give it the name "Optimizer" instead of "Optimizator", expect in the case that you are having the mismatch on purpose to avoid name-clashes. If in doubt, use tools like [Google Ngram](https://books.google.com/ngrams/graph?content=+Optimizator%2C+optimizer&year_start=1800&year_end=2022&corpus=en&smoothing=3) to find the more commonly used term.

File: tests/utility/test_opt_algo.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+location = "tests/helpers/device_parameters/ibm_kyiv/"
+
+
+def level_optimization(level: int, result: list, q: list, qc: list, n: int, psi0: np.array, sim: MrAndersonSimulator):
+    bb = BinaryBackend(n)
+    opt_ = Optimizator(level_opt=level, circ_list=result, qubit_list=q)
+    result_ = opt_.optimize()
+    psi_result_1 = bb.statevector(mp_list=result_ , psi0 = psi0,level_opt= 0, qubit_layout=q)
+    probs_ = np.square(np.absolute(psi_result_1))
```
Comment (romanwixinger) [inline tests/utility/test_opt_algo.py:33]: @paolodr98 What is the idea behind the underscores at the end of the variables?

File: tests/utility/test_opt_algo.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+    """Result 4 """
+
+    goal4 = level_optimization(level= 4, result= result, q= q, qc= qc, n= n, psi0= psi0, sim= sim)
+    print("goal 4: ",goal4)
+    check_04 = np.abs(goal0-goal4)
+    print("check 04: ", check_04)
+    print("")
+
+    assert all((i < 10e-12 for i in check_01)), f"Results from level 0 and 1 mismatch"
```
Comment (romanwixinger) [inline tests/utility/test_opt_algo.py:110]: @paolodr98 Perfect! This is a great unit test!

File: tests/simulation/test_circuit.py [hunk 7]
```diff
@@ -95,3 +106,18 @@ def test_alternative_apply_cnot_three_qubits(BackendClass):
     prop_exp = np.array([0, 0, 0, 0, 0, 0, 0, 1**2])
     assert all(abs(abs(prop_found[i]) - abs(prop_exp[i])) < 0.2 for i in range(8)), \
         f"Propagation failed. Expected prop {prop_exp} vs. found {prop_found}."
+    
+def test_non_close_gate_binary_circuit():
+    n_qubit = 3
+    psi0 = [1] + [0] * (2**n_qubit-1)
+    qubit_layout = [0,1,14]
+    circ = BinaryCircuit(nqubit = n_qubit, depth=1, gates= standard_gates)
```
Comment (romanwixinger) [inline tests/simulation/test_circuit.py:114]: @paolodr98 Do whitespaces before and after "=" in the case it is used with keyword arguments.

File: tests/simulation/test_circuit.py [hunk 8]
```diff
@@ -95,3 +106,18 @@ def test_alternative_apply_cnot_three_qubits(BackendClass):
     prop_exp = np.array([0, 0, 0, 0, 0, 0, 0, 1**2])
     assert all(abs(abs(prop_found[i]) - abs(prop_exp[i])) < 0.2 for i in range(8)), \
         f"Propagation failed. Expected prop {prop_exp} vs. found {prop_found}."
+    
+def test_non_close_gate_binary_circuit():
```
Comment (romanwixinger) [inline tests/simulation/test_circuit.py:110]: @paolodr98 Please remove this function and the whitespaces in case the function is not used.

Comment (romanwixinger) [reply tests/simulation/test_circuit.py:110]: Pro tip: You can use an IDE like Intellj to find errors in the code like spelling errors, unused code, unreachable code, unassigned variables etc.

Comment (romanwixinger) [reply tests/simulation/test_circuit.py:110]: You can probably get a student license for free for another year.

File: tests/simulation/test_circuit.py [hunk 9]
```diff
@@ -51,14 +56,20 @@ def test_alternative_circuit_apply(BackendClass):
...
+            circ.apply(gate=None, i=0)
 
 
 @pytest.mark.parametrize("CircuitClass", circuit_classes)
 def test_circuit_apply_2qubit_gate_raises_ValueError(CircuitClass):
     with pytest.raises(ValueError):
         circ = CircuitClass(nqubit=2, depth=1, gates=standard_gates)
-        circ.apply(gate=helper_gates.CNOT, i=0) # Use apply for 2 qubit (CNOT) gate.
+        if isinstance(circ, BinaryCircuit):
+            circ.update_circuit_list(gate=helper_gates.CNOT, qubit = [0])
```
Comment (romanwixinger) [inline tests/simulation/test_circuit.py:70]: @paolodr98 It seems that the two circuit classes do not have the same interface. Is it necessary that there is a mismatch? Would it be possible to make the interface match with reasonable effort?

Comment (romanwixinger) [reply tests/simulation/test_circuit.py:70]: Just to ensure compatibility

File: tests/simulation/test_backend.py [hunk 10]
```diff
@@ -88,50 +105,68 @@ def test_backends_get_same_result_with_single_qubit_gates(nqubits: int, steps: i
...
         f"The one backend did not generate the same result as the standard backend. Found {psi_one} and {psi_standard}."
+    
+    assert vector_almost_equal(psi_binary, psi_standard, nqubits), \
+        f"The binary backend did not generate the same result as the standard backend. Found {psi_binary} and {psi_standard}."
 
 
 @pytest.mark.parametrize("nqubits, steps", [(n, s) for n in [2, 3, 4, 6, 8, 9, 10] for s in [5]])
 def test_backends_get_same_result_with_random_matrix_products(nqubits, steps):
-    mp_list = generate_random_matrix_products(nqubits, steps=steps)
+    mp_list,mp_list_b = generate_random_matrix_products(nqubits, steps=steps)
```
Comment (romanwixinger) [inline tests/simulation/test_backend.py:137]: @paolodr98 Add a whitespace after the comma in the tuple.

File: tests/simulation/test_backend.py [hunk 11]
```diff
@@ -59,17 +71,22 @@ def test_backend_x(nqubits, Backend):
...
-        mp_list.append([identity for i in range(j)] + [CNOT, 1] + [identity for i in range(nqubits - 2 - j)])
-
     tb = Backend(nqubit=nqubits)
-
-    # Check that it does the correct mapping
     psi0 = np.zeros(2**nqubits)
     psi0[0] = 1.0
-
-    psi1 = tb.statevector(mp_list, psi0)
+    if isinstance(tb, BinaryBackend):
```
Comment (romanwixinger) [inline tests/simulation/test_backend.py:77]: @paolodr98 Let us have chat about the change in the interface between the BinaryBackend and the other backends. It would be great if they were 1:1 replacements of each other.

File: tests/simulation/test_backend.py [hunk 12]
```diff
@@ -13,11 +13,11 @@
 )
 from tests.helpers.functions import (
     vector_almost_equal,
-    generate_random_matrix_products,
```
Comment (romanwixinger) [inline tests/simulation/test_backend.py:16]: @paolodr98 The additional comma serves the following purpose: When you add a new item to the list (a new import), then you wont have to add a comma on the second last line because there is already one. This means that you only have to change on line instead of two which is easier to understand.

File: src/quantum_gates/_simulation/circuit.py [hunk 13]
```diff
@@ -594,6 +594,269 @@ def reset(self):
         self.phi = [0 for i in range(self.nqubit)]
 
 
+class BinaryCircuit(object):
```
Comment (romanwixinger) [inline src/quantum_gates/_simulation/circuit.py:597]: @paolodr98 Could you explain the difference between the different (if any) circuit classes in the next meeting? Maybe we could combine them somehow and remove some redundancy. Now it is a lot of similar code.

File: src/quantum_gates/_simulation/backend.py [hunk 14]
```diff
@@ -212,6 +215,203 @@ def _chunk_list(self, l: list, min_chunk_size: int, optimal_chunk_size: int) ->
...
+        # quantities for the sparse matrix
+        data = []
+        row_indices = []
+        col_indices = []
+
+        for i in range(2**k): # generate only the non zero element in the matrix according to the number of not used qubit
+                k_str = f"{i*(2**k+1):0{2*k}b}"
+                for j in range(2**(2*(m))):
+                    m_str = f"{j:0{2*(m)}b}"
+                    n_str = self.join_str(k_str, m_str, q_n_used, q_used ,k ,m) # merge string
```
Comment (romanwixinger) [inline src/quantum_gates/_simulation/backend.py:325]: @paolodr98 What is the idea behind the join_str method?

**Review by romanwixinger — CHANGES_REQUESTED** [review]
**Feedback**

Thank you for the nice PR! 

I mainly have some inputs regarding the styling and some questions regarding the logic. You will find them in the comments. 

Let us have a discussion next week to go over some of the questions and discuss if we want to keep the interface of BinaryBackend the same as the other backends. I think this would be really helpful but you have to tell me if it is feasible. 

Could you also check the following failed tests and warnings:
- tests/simulation/test_backend.py::test_one_backend_is_faster_than_efficient_backend[17-500] - ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (17,) + inhomogeneous part.
- FAILED tests/simulation/test_backend.py::test_backend_is_faster_than_standard_backend[10-5] - AssertionError: Inserted l with length 5 but expected at least length 8.
- FAILED tests/simulation/test_anderson_simulator.py::test_simulator_speed_for_more_efficient_circuits[2-2] - TypeError: unsupported operand type(s) for -: 'dict' and 'dict'
- quantum-gates/src/quantum_gates/_simulation/simulator.py:443: DeprecationWarning: Treating CircuitInstruction as an iterable is deprecated legacy behavior since Qiskit 1.2, and will be removed in Qiskit 2.0. Instead, use the `operation`, `qubits` and `clbits` named attributes.
    time = data[j][0].duration * dt

File: src/quantum_gates/_simulation/simulator.py [hunk 15]
```diff
@@ -63,44 +64,87 @@ def run(self,
...
+                    if q not in used_q:
+                        used_q.append(q)
+                elif len(x[1]) == 2:
+                    q1 = x[1][0]._index
+                    q2 = x[1][1]._index
+                    if q1 not in used_q:
+                        used_q.append(q1)
+                    if q2 not in used_q:
+                        used_q.append(q2)
+                if x[0].name == 'measure':
```
[the remark this answers is not in the document]
Comment (paolodr98) [reply src/quantum_gates/_simulation/simulator.py:142]: Sure, one index is about the measured qubit and the the second index is about the classical bit where it is stored the information. Store in a list the measured qubits and associated classical bits ensure possible swap in the qubit measurament that are taken in account in the measurament function below.

File: src/quantum_gates/_simulation/simulator.py [hunk 16]
```diff
@@ -63,44 +64,87 @@ def run(self,
...
                 qubits_layout: qubits layout with linear topology (list)
                 psi0: initial state (array)
                 shots: number of realizations (int)
                 device_param: noise and device configurations as dict with the keys specified by DeviceParameters (dict)
                 nqubit: number of qubits used in the circuit, must be compatible with psi0 (int)
+                level_opt (int): Level of optimization for the circuit optimizator for BinaryCircuit only
+
             Returns:
-                  vector of probabilities (array)
+                  dictionary of probabilities: the keys are the binary strings and the values the probabilities (dict)
```
[the remark this answers is not in the document]
Comment (paolodr98) [reply src/quantum_gates/_simulation/simulator.py:82]: Sure, I add an output example in the documentation of the simulator class in the code example. I also add in the doc of the run function a note to remember that the output of the simulator follow the Big Endian order for the bit string.

File: src/quantum_gates/_simulation/simulator.py [hunk 17]
```diff
@@ -294,64 +392,122 @@ def _apply_gates_on_circuit(
...
+
+            if data[j][0].name == 'cx':
+                q_ctr_r = data[j][1][0]._index # index control real qubit
+                q_trg_r = data[j][1][1]._index # index target real qubit
+                q_ctr_v = qubit_layout.index(q_ctr_r) # index control virtual qubit
+                q_trg_v = qubit_layout.index(q_trg_r) # index control virtual qubit
+                for k in range(nqubit):
+                    if k == q_ctr_v:
+                        circ.CNOT(q_ctr_v, q_trg_v, t_int[q_ctr_r][q_trg_r], p_int[q_ctr_r][q_trg_r], p[q_ctr_r], p[q_trg_r], T1[q_ctr_r], T2[q_ctr_r], T1[q_trg_r], T2[q_trg_r])
+                    elif k == q_trg:
```
[the remark this answers is not in the document]
Comment (paolodr98) [reply src/quantum_gates/_simulation/simulator.py:437]: Yeah, I mean q_trg_v as in the case for the 'ecr'

File: src/quantum_gates/_utility/circ_optimizator.py [hunk 18]
```diff
@@ -0,0 +1,491 @@
...
+            return self.circ_list
+
+        if level == 1:
+            result_1 = self.opt_level_1(gate_list=self.circ_list)
+            return result_1
+        
+        elif level == 2:
+            result_1 = self.opt_level_1(gate_list=self.circ_list)
+            result_2 = self.opt_level_2(gate_list=result_1)
+            result_3 = self.opt_level_3(gate_list=result_2)
```
[the remark this answers is not in the document]
Comment (paolodr98) [reply src/quantum_gates/_utility/circ_optimizator.py:80]: Just a cut and paste typo. I delete it

File: src/quantum_gates/_utility/circ_optimizator.py (hunk 1 above)
Comment (paolodr98) [reply src/quantum_gates/_utility/circ_optimizator.py:2]: It's an idea we came up with to speed up the execution of the new BinaryBackend; it takes advantage of the gate patterns in the transpiled circuit. It is not releated with the Qiskit optimization, (I imagined that the name would lead to misunderstandings, but it was the most fitting one at the time I wrote the code). We will write a more detailed documentation

File: tests/utility/test_opt_algo.py (hunk 5 above)
Comment (paolodr98) [reply tests/utility/test_opt_algo.py:33]: Because before defining the `level_optimization` function, these are 'free' variables with a numerical index referring to the optimization level. Then I create the function remove the numerical indexes, but I leave the underscore. I can remove it because there is no real meaning

File: src/quantum_gates/_simulation/backend.py (hunk 14 above)
Comment (paolodr98) [reply src/quantum_gates/_simulation/backend.py:325]: The idea is to calculate only the binary strings for the non-zero elements of the matrix: First, all strings representing the unused qubits are generated and stored in k_str. Then, the strings for the used qubits are generated and stored in m_str. In the `join_str` method, the bit of these strings are placed in the correct position, according to which qubits are used and which are not. I'll provide you an example later during the meeting

**Comment by romanwixinger** [comment]
@paolodr98 Let us try to make the interface of BinaryCircuit compatible with EfficientCircuit such that BinaryCircuit is a drop-in replacement of the EfficientCircuit. Here would be my proposed strategy:

**Add qubit_layout as an optional parameter in the constructor**:

- Modify the BinaryCircuit constructor to include qubit_layout=None, defaulting to list(range(nqubit)). This allows the circuit to store the qubit layout internally, removing the need to pass it to the statevector method. It aligns with EfficientCircuit by simplifying the constructor interface.

**Rename update_circuit_list to apply and adjust its signature:** 

- Rename the method to apply(self, gate: np.ndarray, i: int, j=-1) to match EfficientCircuit. The optional j parameter supports two-qubit gates, defaulting to -1 for single-qubit gates. This ensures consistency across the two circuits.

**Modify gate application methods to use the new apply method**:

- Update methods like I, X, and CNOT to use self.apply(gate, i) for single-qubit gates and self.apply(gate, i, j) for two-qubit gates. This standardizes gate application and simplifies internal logic, ensuring compatibility.

**Update the statevector method to use the internal qubit_layout**:

- Remove qubit_layout and level_opt from the statevector method parameters and use self.qubit_layout internally. This aligns the method with EfficientCircuit, simplifying its usage and making it a 1:1 replacement.

**Ensure attribute consistency**
- Check that attributes like nqubit, gates, and phi are used similarly in both classes. 

I believe that these changes would be enough to make it happen. What do you think? I can also create a new branch to check if the changes are possible.

**Comment by romanwixinger** [comment]
> @paolodr98 Let us try to make the interface of BinaryCircuit compatible with EfficientCircuit such that BinaryCircuit is a drop-in replacement of the EfficientCircuit. Here would be my proposed strategy:
> 
> **Add qubit_layout as an optional parameter in the constructor**:
> 
> * Modify the BinaryCircuit constructor to include qubit_layout=None, defaulting to list(range(nqubit)). This allows the circuit to store the qubit layout internally, removing the need to pass it to the statevector method. It aligns with EfficientCircuit by simplifying the constructor interface.
> 
> **Rename update_circuit_list to apply and adjust its signature:**
> 
> * Rename the method to apply(self, gate: np.ndarray, i: int, j=-1) to match EfficientCircuit. The optional j parameter supports two-qubit gates, defaulting to -1 for single-qubit gates. This ensures consistency across the two circuits.
> 
> **Modify gate application methods to use the new apply method**:
> 
> * Update methods like I, X, and CNOT to use self.apply(gate, i) for single-qubit gates and self.apply(gate, i, j) for two-qubit gates. This standardizes gate application and simplifies internal logic, ensuring compatibility.
> 
> **Update the statevector method to use the internal qubit_layout**:
> 
> * Remove qubit_layout and level_opt from the statevector method parameters and use self.qubit_layout internally. This aligns the method with EfficientCircuit, simplifying its usage and making it a 1:1 replacement.
> 
> **Ensure attribute consistency**
> 
> * Check that attributes like nqubit, gates, and phi are used similarly in both classes.
> 
> I believe that these changes would be enough to make it happen. What do you think? I can also create a new branch to check if the changes are possible.

Link to the branch: https://github.com/CERN-IT-INNOVATION/quantum-gates/tree/circuit-interface

**Comment by paolodr98** [comment]
@romanwixinger I fix the source code and the tests in the new branch to match the interface of the two circuit classes, now should be alright, let me now if there are other things to do.

**Comment by romanwixinger** [comment]
@paolodr98 Great work! I just did some last improvements on the circuit-interface branch. I updated the statevector method of the BinaryBackend to have the qubit layout as optional, such that the class can be used as a 1:1 replacement for the other backends. I also updated the tests, now all relevant tests pass. 

Could you merge the circuit-interface branch into your branch in your repo to update this current PR? As soon as this PR is up to date, I will merge it!

**Comment by paolodr98** [comment]
Done! I merge the circuit-interface branch in the main branch of my fork so now the last commits appear in the PR. Stupidly I left my IBM token on the tutorial notebook, I delete and regenerate it so there is no leakage of sensible information.

**Comment by romanwixinger** [comment]
> Done! I merge the circuit-interface branch in the main branch of my fork so now the last commits appear in the PR. Stupidly I left my IBM token on the tutorial notebook, I delete and regenerate it so there is no leakage of sensible information.

Great job! I would recommend using [python-dotenv](https://dev.to/emma_donery/python-dotenv-keep-your-secrets-safe-4ocn) for handling environment variables like tokens. This way, you do not have to worry about leaking any credentials:)

**Review by romanwixinger — APPROVED** [review]

**Comment by romanwixinger** [comment]
@paolodr98 Could you check why this unit test (test_backends_get_same_result_with_random_matrix_products) fails? There might be an edge case that was not covered.

`python
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/quantum_gates/_simulation/backend.py:265: in statevector
    mp_list_opt = Optimizer(level_opt= 4, circ_list= mp_list, qubit_list=qubit_layout).optimize()
src/quantum_gates/_utility/circ_optimizer.py:96: in optimize
    result_2 = self.opt_level_2(gate_list=result_1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <src.quantum_gates._utility.circ_optimizer.Optimizer object at 0x122668550>
gate_list = [[array([[1, 0, 0, 0],
       [0, 1, 0, 0],
       [0, 0, 0, 1],
       [0, 0, 1, 0]]), [7, 8]], [array([[1.70710678+0.j        , 0.        -0.70710678j],
       [0.        -0.70710678j, 1.70710678+0.j        ]]), [9]]]

    def opt_level_2(self, gate_list: list) -> list:
        """This is the second level of optimization. It watch at the two single qubit gate before and after a two qubit gate:
        If there are one or two single qubit gate before a two qubit gate that act on the same qubit then there are multiplicate together.
        The same is done if there are one or two single qubit gate after the two qubit gate.
        In this way most of the single qubit gate are deleted and the information is stored inside the two qubit gates
    
        Args:
            gate_list (list): List of gate coming from the first level of optimization
    
        Returns:
            list: Return the second level optimize list of gate
        """
    
        q2_gate = 0
        result_2 = []
    
        for i in gate_list: #count the number of two qubit gate in the result
            if len(i[1]) == 2:
                q2_gate += 1
    
        if q2_gate > 0: # check that exists almost one two qubit gate
            for i in range(q2_gate): # iterate over the number of  two qubit gate
    
                # create the snippet
                _snippet = []
                counter = 0
                while len(gate_list[counter][1]) != 2:
    
                    _snippet.append(gate_list[counter])
                    counter += 1
                _snippet.append(gate_list[counter]) # append the 2q gate
                if len(gate_list) > counter+1: # assert that there are gate after the one considered
                    if len(gate_list[counter+1][1]) == 1: # append the right after gate if it is a 1 q gate
                        _snippet.append(gate_list[counter+1])
>                       if len(gate_list[counter+2][1]) == 1: # if it's appended the first closer whatch if append the second after gate if it is a 1 q gate
E                       IndexError: list index out of range

src/quantum_gates/_utility/circ_optimizer.py:175: IndexError
`

**Comment by paolodr98** [comment]
Fix the last problem about the test for the Backend.
